src/Servers/HttpSys/src/UrlPrefixCollection.cs

```diff
@@ -138,12 +146,56 @@ internal void RegisterAllPrefixes(UrlGroup urlGroup)
             {
                 _urlGroup = urlGroup;
                 // go through the uri list and register for each one of them
-                foreach (var pair in _prefixes)
+                // Call ToList to avoid modification when enumerating.
+                foreach (var pair in _prefixes.ToList())
+                {
+                    var urlPrefix = pair.Value;
+                    if (urlPrefix.PortValue == 0)
+                    {
+                        if (urlPrefix.IsHttps)
+                        {
+                            throw new InvalidOperationException("Cannot bind to port 0 with https.");
+                        }
+
+                        if (!FindHttpPortUnsynchronized(pair.Key, urlPrefix))
+                        {
+                            throw new HttpSysException(Marshal.GetLastWin32Error(), "Could not bind to port 0.");
+                        }
+                    }
+                    else
+                    {
+                        // We'll get this index back on each request and use it to look up the prefix to calculate PathBase.
+                        _urlGroup.RegisterPrefix(pair.Value.FullPrefix, pair.Key);
+                    }
+                }
+            }
+        }
+
+        private bool FindHttpPortUnsynchronized(int key, UrlPrefix urlPrefix)
+        {
+            for (var index = BasePortIndex; index < MaxPortIndex; index++)
+            {
+                try
+                {
+                    // Bit of complicated math to always try 3000 ports, starting from NextPortIndex + 5000,
+                    // circling back around if we go above 8000 back to 5000, and so on.
+                    var port = ((index + NextPortIndex) % MaxPortIndex) + BasePort;
```

Comment:
You're using MaxPortIndex in two unrelated ways:
1) how many reties to do. 1000 would be more than adequate. 
2) what's the upper port range. This is an arbitrary figure that we could bump to the tcp max port of 64k.